### PR TITLE
 More permanent fix AnimeDigitalNetwork

### LINKED ISSUE #3832: [Presence | Bug] Anime Digital Network
**Describe the Bug**
The status just show "Anime Digital Network" when watching a episode

**Steps to Reproduce**
1. Open https://animedigitalnetwork.fr/
2. Start an episode

**Expected Behavior**
The status should be updated with anime title, episode...

**Screenshots**
unknown (2)

**Environment**
- OS: Windows 10
- Browser: Chrome
- Version: 91.0.4472.77 (64-bit)

### PR DESCRIPTION
**Pull request by DarkVillager**
Resolves #3832
Metadata.json:
- Bumped version (big fix).
- Added new tag.
Presence.ts:
- Added new way to get title.
- Added that it shows the title & episode number even if it isnt playing
- Fixed all bugs

image
image
image
image
image

### CODE REVIEW

**Comment by SlowLife1661** [comment]
fix indent

**Review by Slowlife01 — DISMISSED** [review]

File: websites/A/AnimeDigitalNetwork/presence.ts [hunk 1]
```diff
@@ -32,18 +30,24 @@ presence.on("UpdateData", async () => {
...
+    const catalogue = document.querySelector("#root > div > div > div.sc-pkSvE.kPCOPp > div > div > div.sc-AxjAm.khAjwj.sc-psDXd.iazofB > div > h2 > span"); 
+    if (catalogue) 
+      presenceData.details = "Browsing...";
+    else {
+      const episode = JSON.parse(document.querySelector("#root > div > div > div.sc-pkSvE.kPCOPp > div > div > div.sc-psOyd.fIwdpb > script").textContent);
+      presenceData.details = "Looking at";
+      presenceData.state = episode.name;
+    }
   } else presenceData.details = "Browsing...";
   if (presenceData.details === null) {
```
Comment (SlowLife1661) [inline websites/A/AnimeDigitalNetwork/presence.ts:47]:
```suggestion
  if (!presenceData.details) {
```

**Review by Slowlife01 — DISMISSED** [review]

File: websites/A/AnimeDigitalNetwork/presence.ts [hunk 2]
```diff
@@ -11,17 +11,15 @@ presence.on("UpdateData", async () => {
...
-      const title = document.querySelector(
-          "#root > div > div > div.sc-pbWVv.hTvDIL > div > div:nth-child(1) > div.sc-kbKFCX.sc-kgMcbC.kecmmo > div:nth-child(1) > div > div > h1 > a"
-        ),
-        timestamps = presence.getTimestamps(
+    const episode = JSON.parse(document.querySelector("#root > div > div > div.sc-pkSvE.kPCOPp > div > div:nth-child(1) > script").textContent);
+    if (!isNaN(video.duration)) {
+      const timestamps = presence.getTimestamps(
           Math.floor(video.currentTime),
           Math.floor(video.duration)
         );
```
Comment (Bas950) [inline websites/A/AnimeDigitalNetwork/presence.ts:21]: might as well use presence.getMediaTimestamps or it's called something like that.

**Review by Bas950 — CHANGES_REQUESTED** [review]

File: websites/A/AnimeDigitalNetwork/presence.ts [hunk 3]
```diff
@@ -13,37 +13,78 @@ presence.on("UpdateData", async () => {
...
+      const timestamps = presence.getTimestampsfromMedia(video);
+      presenceData.details = episode.partOfSeries.name;
       presenceData.smallImageKey = video.paused ? "pause" : "play";
       presenceData.smallImageText = video.paused
         ? (await strings).pause
         : (await strings).play;
       presenceData.endTimestamp = timestamps[1];
+      presenceData.buttons = [
+        {
+          label: "Watch",
```
Comment (Bas950) [inline websites/A/AnimeDigitalNetwork/presence.ts:31]:
```suggestion
          label: "Watch Episode",
```

File: websites/A/AnimeDigitalNetwork/presence.ts [hunk 4]
```diff
@@ -13,37 +13,78 @@ presence.on("UpdateData", async () => {
...
       if (video.paused) {
         delete presenceData.startTimestamp;
         delete presenceData.endTimestamp;
       }
+    } else {
+      presenceData.details = "Looking at";
+      presenceData.state = episode.partOfSeries.name;
+      presenceData.buttons = [
+        {
+          label: "Check Out",
```
Comment (Bas950) [inline websites/A/AnimeDigitalNetwork/presence.ts:45]:
```suggestion
          label: "View Page",
```

File: websites/A/AnimeDigitalNetwork/presence.ts [hunk 5]
```diff
@@ -13,37 +13,78 @@ presence.on("UpdateData", async () => {
...
-    presence.setTrayTitle();
-    presence.setActivity();
+    if (catalogue) {
+      presenceData.details = "Browsing...";
+      presenceData.buttons = [
+        {
+          label: "Browse",
+          url: document.location.href
+        }
+      ];
```
Comment (Bas950) [inline websites/A/AnimeDigitalNetwork/presence.ts:61]:
```suggestion
```

File: websites/A/AnimeDigitalNetwork/presence.ts [hunk 6]
```diff
@@ -13,37 +13,78 @@ presence.on("UpdateData", async () => {
...
+      const episode = JSON.parse(
+        document.querySelector(
+          "#root > div > div > div.sc-pkSvE.kPCOPp > div > div > div.sc-psOyd.fIwdpb > script"
+        ).textContent
+      );
+      presenceData.details = "Looking at";
+      presenceData.state = episode.name;
+      presenceData.buttons = [
+        {
+          label: "Check Out",
```
Comment (Bas950) [inline websites/A/AnimeDigitalNetwork/presence.ts:72]:
```suggestion
          label: "View Page",
```

File: websites/A/AnimeDigitalNetwork/presence.ts [hunk 7]
```diff
@@ -13,37 +13,78 @@ presence.on("UpdateData", async () => {
...
+    }
   } else {
-    presence.setActivity(presenceData);
+    presenceData.details = "Browsing...";
+    presenceData.buttons = [
+      {
+        label: "Browse",
+        url: document.location.href
+      }
+    ];
```
Comment (Bas950) [inline websites/A/AnimeDigitalNetwork/presence.ts:84]:
```suggestion
```

**Review by Bas950 — CHANGES_REQUESTED** [review]
Add a setting to toggle buttons

File: websites/A/AnimeDigitalNetwork/presence.ts [hunk 8]
```diff
@@ -10,40 +10,76 @@ presence.on("UpdateData", async () => {
...
+    }
   } else {
-    presence.setActivity(presenceData);
+    presenceData.details = "Browsing...";
+
+    if (!presenceData.details) {
+      presence.setTrayTitle();
+      presence.setActivity();
+    } else presence.setActivity(presenceData);
   }
```
Comment (Bas950) [inline websites/A/AnimeDigitalNetwork/presence.ts:84]: are you sure this should be in an else?

**Comment by Bas950** [comment]
Upload new proof because of last commit.

**Comment by DarkVillager** [comment]
image

**Comment by Bas950** [comment]
One that doesn't just say browsing...

File: websites/A/AnimeDigitalNetwork/presence.ts [hunk 9]
```diff
@@ -10,40 +10,74 @@ presence.on("UpdateData", async () => {
...
-          "#root > div > div > div.sc-pbWVv.hTvDIL > div > div:nth-child(1) > div.sc-kbKFCX.sc-kgMcbC.kecmmo > div:nth-child(1) > div > div > h1 > a"
-        ),
-        timestamps = presence.getTimestamps(
-          Math.floor(video.currentTime),
-          Math.floor(video.duration)
-        );
-      presenceData.details = title.textContent;
+    const episode = JSON.parse(
+      document.querySelector(
+        "#root > div > div > div.sc-pkSvE.kPCOPp > div > div:nth-child(1) > script"
```
Comment (Slowlife01) [inline websites/A/AnimeDigitalNetwork/presence.ts:19]: How about using instead? And there are many more other ways to get this JSON
```js
document.querySelector('[type="application/ld+json"]')
```

File: websites/A/AnimeDigitalNetwork/presence.ts [hunk 10]
```diff
@@ -10,40 +10,74 @@ presence.on("UpdateData", async () => {
...
+    const catalogue = document.querySelector(
+      "#root > div > div > div.sc-pkSvE.kPCOPp > div > div > div.sc-AxjAm.khAjwj.sc-psDXd.iazofB > div > h2 > span"
     );
-    presenceData.details = "Looking at";
-    presenceData.state = title.textContent;
+    if (catalogue) presenceData.details = "Browsing...";
+    else {
+      const episode = JSON.parse(
+        document.querySelector(
+          "#root > div > div > div.sc-qPmLO.gspUft > div > div > div.sc-qXiOn.iVbXhX > script"
```
Comment (Slowlife01) [inline websites/A/AnimeDigitalNetwork/presence.ts:63]: Same with above

File: websites/A/AnimeDigitalNetwork/presence.ts [hunk 11]
```diff
@@ -60,7 +60,7 @@ presence.on("UpdateData", async () => {
     else {
       const episode = JSON.parse(
         document.querySelector(
-          "#root > div > div > div.sc-pkSvE.kPCOPp > div > div > div.sc-psOyd.fIwdpb > script"
+          "#root > div > div > div.sc-qPmLO.gspUft > div > div > div.sc-qXiOn.iVbXhX > script"
```
Comment (Bas950) [inline websites/A/AnimeDigitalNetwork/presence.ts:63]: Shouldnt you use nth:child or something that is more against the class name changes.

**Comment by DarkVillager** [comment]
I can’t work on this till Monday just to let the mods know.

**Review by Bas950 — APPROVED** [review]

**Review by Slowlife01 — APPROVED** [review]

**Pull request merged**